src/leopard_em/pydantic_models/orientation_search.py

```diff
@@ -146,23 +178,105 @@ def euler_angles_offsets(self) -> torch.Tensor:
         if not self.enabled:
             return torch.zeros((1, 3))
 
-        psi_values = torch.arange(
-            -self.in_plane_angular_step_coarse,
-            self.in_plane_angular_step_coarse + EPS,
-            self.in_plane_angular_step_fine,
-        )
-        theta_values = torch.arange(
-            0.0,
-            self.out_of_plane_angular_step_coarse + EPS,
-            self.out_of_plane_angular_step_fine,
+        return get_local_high_resolution_angles(
+            coarse_in_plane_step=self.in_plane_angular_step_coarse,
+            coarse_out_of_plane_step=self.out_of_plane_angular_step_coarse,
+            fine_in_plane_step=self.in_plane_angular_step_fine,
+            fine_out_of_plane_step=self.out_of_plane_angular_step_fine,
+            base_grid_method=self.base_grid_method,
         )
-        phi_values = torch.arange(
-            -self.out_of_plane_angular_step_coarse,
-            self.out_of_plane_angular_step_coarse + EPS,
-            self.out_of_plane_angular_step_fine,
+
+
+class ConstrainedOrientationConfig(BaseModel2DTM):
+    """Serialization and validation of constrained orientation parameters.
+
+    Attributes
+    ----------
+    enabled: bool
+        Whether to enable constrained orientation search.
+    in_plane_step: float
+        Angular step size for in-plane rotations in degrees.
+        Must be greater than or equal to 0.
+    out_of_plane_step: float
+        Angular step size for out-of-plane rotations in degrees.
+        Must be greater than or equal to 0.
+    rotation_axis_euler_angles: list[float]
+        List of Euler angles (phi, theta, psi) for the rotation axis.
+    phi_min: float
+        Minimum value for the phi angle in degrees.
+    phi_max: float
+        Maximum value for the phi angle in degrees.
+    theta_min: float
+        Minimum value for the theta angle in degrees.
+    theta_max: float
+        Maximum value for the theta angle in degrees.
+    psi_min: float
+        Minimum value for the psi angle in degrees.
+    psi_max: float
+        Maximum value for the psi angle in degrees.
+    """
+
+    enabled: bool = True
+    in_plane_step: float = 1.5
+    out_of_plane_step: float = 2.5
+    rotation_axis_euler_angles: list[float] = Field(default=[0.0, 0.0, 0.0])
+    phi_min: float = 0.0
+    phi_max: float = 360.0
+    theta_min: float = 0.0
+    theta_max: float = 180.0
+    psi_min: float = 0.0
+    psi_max: float = 360.0
+    base_grid_method: Literal["uniform", "healpix", "basic"] = "uniform"
+
+    @property
+    def euler_angles_offsets(self) -> torch.Tensor:
```

Comment:
Update return type to `tuple[torch.Tensor, torch.Tensor]` and the docstring accordingly for the offsets and rotated Euler angles